libcontainer/nsenter/nsexec.c

```diff
@@ -569,6 +587,182 @@ void join_namespaces(char *nslist)
 /* Defined in cloned_binary.c. */
 extern int ensure_cloned_binary(void);
 
+void receive_fd(int sockfd, int new_fd)
+{
+	int bytes_read;
+	struct msghdr msg = {0,};
+	struct cmsghdr *cmsg;
+	struct iovec iov = {0,};
+	char null_byte = '\0';
+	int ret;
+	int *fd_payload;
+
+	iov.iov_base = &null_byte;
+	iov.iov_len = 1;
+
+	msg.msg_iov = &iov;
+	msg.msg_iovlen = 1;
+
+	msg.msg_controllen = CMSG_SPACE(sizeof(int));
+	msg.msg_control = alloca(msg.msg_controllen);
+	memset(msg.msg_control, 0, msg.msg_controllen);
+
+	msg.msg_controllen = CMSG_LEN(1 * sizeof(int));
+
+	bytes_read = recvmsg(sockfd, &msg, 0);
+	if (bytes_read != 1)
+		bail("failed to receive fd from unix socket %d", sockfd);
+	if (msg.msg_flags & MSG_CTRUNC)
+		bail("received truncated control message from unix socket %d", sockfd);
+
+	cmsg = CMSG_FIRSTHDR(&msg);
+	if (!cmsg)
+		bail("received message from unix socket %d without control message", sockfd);
+
+	if (cmsg->cmsg_level != SOL_SOCKET)
+		bail("received unknown control message from unix socket %d: cmsg_level=%d", sockfd, cmsg->cmsg_level);
+
+	if (cmsg->cmsg_type != SCM_RIGHTS)
+		bail("received unknown control message from unix socket %d: cmsg_type=%d", sockfd, cmsg->cmsg_type);
+
+	fd_payload = (int *) CMSG_DATA (cmsg);
+	if ((cmsg->cmsg_len - CMSG_LEN (0)) != sizeof(int))
+		bail("received control message from unix socket %d with too many fds: %d", sockfd);
+
+	ret = dup2(*fd_payload, new_fd);
+	if (ret < 0)
+		bail("cannot dup2 fd %d to %d", *fd_payload, new_fd);
+
+	ret = close(*fd_payload);
+	if (ret < 0)
+		bail("cannot close fd %d", *fd_payload);
+}
+
+void send_fd(int sockfd, int fd)
+{
+	int bytes_written;
+	struct msghdr msg = {0,};
+	struct cmsghdr *cmsg;
+	struct iovec iov[1] = {0,};
+	char null_byte = '\0';
+
+	iov[0].iov_base = &null_byte;
+	iov[0].iov_len = 1;
+
+	msg.msg_iov = iov;
+	msg.msg_iovlen = 1;
+
+	msg.msg_controllen = CMSG_SPACE(sizeof(int));
+	msg.msg_control = alloca(msg.msg_controllen);
+	memset(msg.msg_control, 0, msg.msg_controllen);
+
+	cmsg = CMSG_FIRSTHDR(&msg);
+	cmsg->cmsg_level = SOL_SOCKET;
+	cmsg->cmsg_type = SCM_RIGHTS;
+	cmsg->cmsg_len = CMSG_LEN(sizeof(int));
+	memcpy(CMSG_DATA(cmsg), &fd, sizeof(int));
```

Comment:
maybe just 
```c
*(int *)CMSG_DATA(cmsg) = fd
```
?

Reply:
I changed this to use memcpy instead. That cast creates a warning on centos:

```
nsexec.c:733:2: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
  *(int *)CMSG_DATA(cmsg) = fd;
```

I changed it to a memcpy, as man cmsg suggest also.